### PR TITLE
Remove nested tensor dispatch key if the op is CompositeImplicit

### PR DESCRIPTION
**Pull request by YifanShenSZ**
Historically, nested tensor dispatch key was left out from composite, as pointed out by issue [#79048](https://github.com/pytorch/pytorch/issues/79048). As a result, if an op is declared to be CompositeImplicit (e.g. dropout, softmax, linear), then when calling on nested tensor the dispatcher would raise "not registered" error rather than call into the CompositeImplicit op.

The hacky solution was to add "NestedTensorCPU, NestedTensorCUDA" entry to "native_functions.yaml" even if the op is CompositeImplicit. This is flaky, so it may cause some unexpected trouble.

As pull request [#80273](https://github.com/pytorch/pytorch/pull/80273) resolves the historical problem by adding AutogradNestedTensor key to CompositeImplicit, we can now remove the hacky practice.

In addition, the `torch.inference_mode()` constraint can be removed. It was flaky in the past:
* some ops could be called without `torch.inference_mode()`, but some must otherwise they would get dispatched to regular tensor kernels rather than the desired nested tensor ones
* if created a nested tensor out of a `torch.inference_mode()` scope, then passing it to an `@torch.inference_mode()` decorated function would trigger "try to set version counter in inference mode" error

Now the CompositeImplicit nested tensor dispatching behaviour under  `torch.inference_mode()` is to get "not registered" error as expected, since CompositeImplicit includes only AutogradNestedTensor key, no NestedTensorCPU nor NestedTensorCUDA. Maybe @drisspg can add them in the future.

### CODE REVIEW

File: aten/src/ATen/native/Dropout.cpp [hunk 1]
```diff
@@ -117,6 +117,9 @@ Tensor native_dropout_backward_cpu(const Tensor& grad, const Tensor& mask, doubl
 Tensor dropout(const Tensor& input, double p, bool train) {
   auto result = [&]() {
     NoNamesGuard guard;
+    if (input.is_nested()) {
```
Comment (jbschlosser) [inline aten/src/ATen/native/Dropout.cpp:120]: As discussed offline, I think we should avoid `is_nested()` calls whenever possible, letting the dispatcher do the work. Given that, I'd suggest leaving dropout in its current state of having both `CompositeImplicitAutograd` and `NestedTensor*` keys registered to it (and similarly for linear).

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -4435,9 +4431,6 @@
 # softmax allows positional dtype, unlike most operators, because kwonly is BC-breaking when loading jit models.
 - func: softmax.int(Tensor self, int dim, ScalarType? dtype=None) -> Tensor
   variants: function, method
-  dispatch:
-    CompositeImplicitAutograd: softmax
-    NestedTensorCPU, NestedTensorCUDA: softmax
```
Comment (jbschlosser) [inline aten/src/ATen/native/native_functions.yaml:4440]: I'm good with this softmax change :) Now that nested tensor can be treated as composite implicit, this properly calls `softmax`, which redispatches to `_softmax`, which has special NT registered kernels for it.
